src/main/java/org/gridsuite/shortcircuit/server/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,229 @@
+                    resultService.delete(resultContext.getResultUuid());
+                }
+            } finally {
+                futures.remove(resultContext.getResultUuid());
+                cancelComputationRequests.remove(resultContext.getResultUuid());
+            }
+        };
+    }
+
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> cancelAsync(CancelContext.fromMessage(message));
+    }
+
+    protected abstract void saveResult(Network network, AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void sendResultMessage(AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void publishFail(AbstractResultContext<R> resultContext, String message);
+
+    /**
+     * Do some extra task before running the computation, e.g. print log or init extra data for the run context
+     * @param ignoredRunContext This context may be used for further computation in overriding classes
+     */
+    protected void preRun(R ignoredRunContext) {
+        LOGGER.info("Run {} computation ...", getComputationType());
+    }
+
+    protected S run(Network network, R runContext, UUID resultUuid) throws Exception {
+        String provider = runContext.getProvider();
+        AtomicReference<ReportNode> rootReporter = new AtomicReference<>(ReportNode.NO_OP);
+        ReportNode reportNode = ReportNode.NO_OP;
+
+        if (runContext.getReportInfos() != null && runContext.getReportInfos().reportUuid() != null) {
+            final String reportType = runContext.getReportInfos().computationType();
+            String rootReporterId = runContext.getReportInfos().reporterId() == null ? reportType : runContext.getReportInfos().reporterId() + "@" + reportType;
+            rootReporter.set(ReportNode.newRootReportNode().withMessageTemplate(rootReporterId, rootReporterId).build());
+            reportNode = rootReporter.get().newReportNode().withMessageTemplate(reportType, String.format("%s (%s)", reportType, provider))
+                    .withUntypedValue("providerToUse", provider).add();
+            // Delete any previous computation logs
+            observer.observe("report.delete",
+                    runContext, () -> reportService.deleteReport(runContext.getReportInfos().reportUuid(), reportType));
+        }
+        runContext.setReportNode(reportNode);
+
+        preRun(runContext);
+        CompletableFuture<S> future = runAsync(network, runContext, provider, resultUuid);
+        S result = future == null ? null : observer.observeRun("run", runContext, future::get);
+        postRun(runContext, rootReporter);
+        return result;
+    }
+
+    /**
+     * Do some extra task after running the computation
+     * @param ignoredRunContext This context may be used for extra task in overriding classes
+     * @param rootReportNode
+     */
+    protected void postRun(R ignoredRunContext, AtomicReference<ReportNode> rootReportNode) { }
```

Comment:
In my opinion, we should implement the default behaviour to avoid the implementation in others children classes

Reply:
done